File: llm/interface.ts
import { loadServiceConfig } from "./config.js";
import { ProviderRegistry } from "./providers/provider-registry.js";
import { tryProviderWithFallback, FallbackExhaustedError } from "./providers/fallback.js";
import type {
  LlmRequest, LlmResponse, LlmMeta, LlmOptions, LlmHandle, FallbackStep,
  StreamChunk,
} from "./protocol.js";

export type {
  LlmRequest, LlmResponse, LlmMeta, LlmOptions, LlmHandle, FallbackStep,
  Message, ToolDefinition, ToolCall, TokenUsage, StreamChunk,
} from "./protocol.js";
export { loadServiceConfig } from "./config.js";

export function createLlm(options?: LlmOptions): LlmHandle {
  const config = loadServiceConfig(options?.configPath);
  const registry = new ProviderRegistry(config);

  const sessionOverrides = new Map<string, ReturnType<typeof registry.getPrimary>>();

  return {
    async send(req: LlmRequest): Promise<LlmResponse> {
      const start = Date.now();

      const override = sessionOverrides.get(req.sessionId);
      const candidates = override ? [override] : registry.getCandidates();

      const result = await tryProviderWithFallback(
        candidates,
        (candidate) => registry.getAdapter(candidate.providerId).sendMessage({
          messages: req.messages,
          model: candidate.model,
          tools: req.tools,
          maxTokens: req.maxTokens,
        }),
      );

      const durationMs = Date.now() - start;
      const { usedCandidate, attempts } = result;

      const failedAttempts = attempts.filter((a) => a.reason !== null);
      const fallbackChain: FallbackStep[] | undefined =
        failedAttempts.length > 0
          ? attempts.map((a) => ({
              providerId: a.providerId,
              model: a.model,
              reason: a.reason,
              durationMs: a.durationMs,
            }))
          : undefined;

      if (failedAttempts.length > 0) {
        sessionOverrides.set(req.sessionId, usedCandidate);
      }

      return {
        reply: result.data.reply,
        toolCalls: result.data.toolCalls,
        usage: result.data.usage,
        meta: {
          sessionId: req.sessionId,
          agentId: req.agentId,
          providerId: usedCandidate.providerId,
          model: usedCandidate.model,
          durationMs,
          fallbackChain,
        },
      };
    },

    async *sendStream(req: LlmRequest): AsyncIterable<StreamChunk> {
      const start = Date.now();
      const override = sessionOverrides.get(req.sessionId);
      const candidates = override ? [override] : registry.getCandidates();

      let lockedCandidate: ReturnType<typeof registry.getPrimary> | null = null;

      for (const candidate of candidates) {
        try {
          const stream = registry.getAdapter(candidate.providerId).sendMessageStream({
            messages: req.messages,
            model: candidate.model,
            tools: req.tools,
            maxTokens: req.maxTokens,
          });

          // 消费流：连接阶段错误在这里捕获并 fallback
          let hasYielded = false;

          for await (const chunk of stream) {
            if (!hasYielded) {
              hasYielded = true;
              lockedCandidate = candidate;
            }
            yield chunk;
          }

          // 流正常结束
          return;
        } catch (err) {
          // 如果已经 yield 过 chunks，不能 fallback，直接抛出
          if (lockedCandidate) throw err;

          // 连接阶段失败，继续尝试下一个 candidate
          // 静默处理，由外层控制是否继续
        }
      }

      // 所有 candidate 都失败了
      throw new FallbackExhaustedError(
        candidates.map((c) => ({
          providerId: c.providerId,
          model: c.model,
          reason: "unknown",
          permanence: "transient",
          durationMs: Date.now() - start,
          retriesUsed: 0,
          timestamp: new Date().toISOString(),
        })),
      );
    },
  };
}

File: src/main/scala/nebscala/llm/interface.scala
package nebscala.llm

import nebscala.shared.{LlmHandle, LlmRequest, LlmResponse, StreamChunk, FallbackStep, LlmOptions, LlmMeta, TokenUsage}
import cats.effect.IO
import cats.syntax.all.*

object LlmInterface:
  def createLlm(options: Option[LlmOptions] = None): LlmHandle[IO] =
    val config = Config.loadServiceConfig(options.flatMap(_.configPath))
    val registry = ProviderRegistry(config)
    val sessionOverrides = scala.collection.mutable.Map.empty[String, ModelCandidate]

    new LlmHandle[IO]:
      def send(req: LlmRequest): IO[LlmResponse] =
        val start = System.currentTimeMillis()
        val candidates = sessionOverrides.get(req.sessionId).toList ++ registry.getCandidates()
          .filterNot(c => sessionOverrides.get(req.sessionId).exists(o => o.providerId == c.providerId && o.model == c.model))

        Fallback.tryProviderWithFallback[AdapterResponse](
          candidates,
          candidate => registry.getAdapter(candidate.providerId).sendMessage(
            SendMessageParams(req.messages, candidate.model, req.tools, req.maxTokens)
          )
        ).map { result =>
          val durationMs = System.currentTimeMillis() - start
          val failedAttempts = result.attempts.filter(_.reason.isDefined)
          val fallbackChain = if failedAttempts.nonEmpty then
            Some(result.attempts.map(a => FallbackStep(a.providerId, a.model, a.reason.map(_.toString), a.durationMs)))
          else None

          if failedAttempts.nonEmpty then
            sessionOverrides(req.sessionId) = result.usedCandidate

          LlmResponse(
            reply = result.data.reply,
            toolCalls = result.data.toolCalls,
            usage = result.data.usage,
            meta = LlmMeta(
              sessionId = req.sessionId,
              agentId = req.agentId,
              providerId = result.usedCandidate.providerId,
              model = result.usedCandidate.model,
              durationMs = durationMs,
              fallbackChain = fallbackChain
            )
          )
        }

      def sendStream(req: LlmRequest): fs2.Stream[IO, StreamChunk] =
        val candidates = sessionOverrides.get(req.sessionId).toList ++ registry.getCandidates()
          .filterNot(c => sessionOverrides.get(req.sessionId).exists(o => o.providerId == c.providerId && o.model == c.model))

        fs2.Stream.eval(IO.ref(false)).flatMap { lockedRef =>
          fs2.Stream.eval(IO.ref(List.empty[FallbackAttempt])).flatMap { failureRef =>
            def tryCandidate(candidates: List[ModelCandidate]): fs2.Stream[IO, StreamChunk] =
              candidates match
                case Nil =>
                  fs2.Stream.eval(
                    failureRef.get.flatMap { failures =>
                      IO.raiseError(new FallbackExhaustedError(
                        if failures.nonEmpty then failures
                        else candidates.map(c => FallbackAttempt(c.providerId, c.model, None, None, 0, 0, java.time.Instant.now().toString))
                      ))
                    }
                  )
                case candidate :: rest =>
                  val stream = registry.getAdapter(candidate.providerId).sendMessageStream(
                    SendMessageParams(req.messages, candidate.model, req.tools, req.maxTokens)
                  )
                  stream.evalTap { chunk =>
                    chunk match
                      case StreamChunk.TextDelta(_) | StreamChunk.ToolCallChunk(_) =>
                        lockedRef.set(true)
                      case _ => IO.unit
                  }.handleErrorWith { err =>
                    fs2.Stream.eval(lockedRef.get).flatMap { locked =>
                      if locked then
                        fs2.Stream.eval(IO.raiseError(err))
                      else
                        val attempt = FallbackAttempt(
                          candidate.providerId, candidate.model,
                          Some(Fallback.classifyError(err).reason),
                          Some(Fallback.classifyError(err).permanence),
                          0, 0, java.time.Instant.now().toString
                        )
                        fs2.Stream.eval(failureRef.update(_ :+ attempt)).flatMap(_ =>
                          tryCandidate(rest))
                    }
                  }

            tryCandidate(candidates)
          }
        }
